### ISSUE TITLE
Submodule not checking out for older versions

### ISSUE THREAD

**Issue by ghost** [haskell-nix/hnix#536]
A project we're building has an older version of hnix as a transitive dependency, and doesn't build since a few days ago, because the `data/nix` submodule cannot be checked out. Apparently, the targeted commit is not part of the `haskell-nix/nix` repo anymore. This happens with all (most?) commits that are older than [PR 530](https://github.com/haskell-nix/hnix/pull/530).

It looks like the `haskell-nix/nix` commits that have been targeted by earlier versions are not accessible anymore, and can't even be checked out manually. This affects at least these two commits:
- `b18a4898ff990c0b8ec1d3571d23a054a7d4c5cd` 
- `c43033101cbba5f23e8b1165521a8a6e1eb082ea`

This could be related to whatever (a force-push?) resolved #533 @jwiegley?

**Comment by tfc** [comment]
I just ran into this problem. It would be awesome if commits didn't go lost.

**Comment by Anton-Latukha** [comment]
Looked, I've myself rebased all branches I had to a new master, so that commit in my local repo is no more.
I would work toward a new release.

You and I can find the state that we need in one of 93 forks that were made.

**Comment by Anton-Latukha** [comment]
And then I would look to keep a branch up for this case.

**Comment by Anton-Latukha** [comment]
I've pocked around GitHub, the search shows that those commits, particularly referenced in head post, are not on GitHub, seems like forks missed recent commits, and I've just mindlessly rebased all my stuff.

If someone from @haskell-nix has those commits (just before Jan 9), please, check out to them and form a branch to upload it. Thank you.

**Comment by ghost** [comment]
>If someone from @haskell-nix has those commits (just before Jan 9), please, check out to them and form a branch to upload it. Thank you.

These commits still seem to be lingering around on github itself:

https://github.com/haskell-nix/nix/commit/c43033101cbba5f23e8b1165521a8a6e1eb082ea

https://github.com/haskell-nix/nix/commit/b18a4898ff990c0b8ec1d3571d23a054a7d4c5cd

I'm speculating here, but it might be enough to simply [put a tag](https://github.com/haskell-nix/nix/tags) on these existing commits, to make them once again available.

**Comment by Anton-Latukha** [comment]
Oh, wait, yes.
Thank you.

You drawn my attention to that this hashes referenced are not HNix, but Nix hashes. No wonder my search on HNix found nothing.

---

tldr; See the last chapter: "Solution".

---
<sub>Happens people ask and suspect why I belabor. Thinking something strange of me.</sub>

<sub>I write like this not only to be clear, but to also expand both your and my and all others understandings at the same time and remember things</sub>

---

#### Let's together understand the Git better

On the fresh, while I understood the whole situation and freshly touched the topic - I would write down Git knowledge.

##### The situation

We see that commits are still in GitHub cache, but they are not only in a detached state - they are not present in the trees of the `haskell-nix/nix`. Cloned `haskell-nix/nix` does not contain those commits. Hence, thank you Git & GitHub, - they are still lingering as the objects inside `.git/modules/data/nix` of the HNix repo itself for us.

Git clones the branch tree graph. And these commits are not in the branch graph.

So when anything clones the `haskell-nix/nix` and tries to refer to this commits - commits are not there. But they are still inside cloned `haskell-nix/hnix, and in `.git` inside of the remote `haskell-nix/nix` that GitHub stores, that is why it exposes them, but they can not be obtained through Git into the local repo.

##### How to do properly

The proper way here is to create an archived branch in `haskell-nix/nix` and push it, to have the old version of the state, as soon as there is a conflict during rebase, and then proceed with the rebase.

This behavior would allow to solve a lot upfront. And allow to easier solve or ease-out the sync of the fork to the upstream,  if rebase is difficult, or if you need to rewrite of the branch that someone is also tracking. If you would rewrite history - at least there would be the old state for everyone, you included to rely on.

---

##### About tags

1. One must have access rights to the repository to put tags there, there are no other ways.
  It is not possible to open a pull request to create a tag. If the tags had been included in the PRs - we would have the cases when forks merge their tags into upstream, and then we should be sorting them out all the time in the PRs, which is not the case.

2. Even if git tag is fully promoted, into annotated (which creates the full proper git object for the entity) and GPG signed - it is still not possible to make a pull request for it.

3. We need to get commits into the `haskell/nix`, they are not there. For tag to work the commit it attached to must be in the git branches graph to be pushed, which can be seen from that we push into remote git branches only. So, the tag can not be pushed into remote without his attached commit, which also should be a part of some branch.

##### About branching

On the contrary, the branches are for free in git, and also the main tool to solve things, all docs remind us of that. The branch definitely works in this case.

##### Main conclusion from all this

Always backup the old state into a branch. Then do rewriting rebase. If the old tree which got rewritten possibly in use by someone - push the branch into the remote repo. Branches are free and the main Git entity, so just close your eyes on the branch list, do not care about how much branches you have, and you would have much better Git experience and tidier Git history also.

##### Solution

PR https://github.com/haskell-nix/nix/pull/3

**Comment by ghost** [comment]
Thank you for sorting this out, much appreciated!

>Always backup the old state into a branch. Then do rewriting rebase. If the old tree which got rewritten possibly in use by someone - push the branch into the remote repo. 

To expand on that, you should always treat published branches as "in use by someone", especially if you're working in an environment like nix, where reproducibility is important, and targeting specific commits of a repository is the usual way to achieve this.

A better approach could be to *first* switch into a new branch, do the rebasing there, and publish it in the new branch afterwards.

> PR haskell-nix/nix#3

What are the specific steps necessary for resolving this at last? 
Just merge the PR? (Would that break something on the master branch over there?)

Do we need to ping someone to make them aware of this?

**Comment by Anton-Latukha** [comment]
As I said I wrote so much to completely understand everything myself.

And documented that.

The solution is simple - find & reintroduce the commits into the repo, store them into an archived branch.

Everything is said in the PR. I also just redacted it.

**Comment by Anton-Latukha** [comment]
I would say that the proper way is to automate such behavior to further not care about it.

Write a script and put it into `~/.config/git/config`

I've just opened my: https://github.com/Anton-Latukha/config-git to the public. It synergizes with https://github.com/Anton-Latukha/fish-functions.

There are a couple of things. & `commit-staged*` is currently broken, it does the opposite thing to what is expected.

In the repo there is a notable smarter case of the "git commit hook" then you would find over the Internet elsewhere.
  * It is portable over machines and repositories, it requires only symbolic link into the default place.
  * Hook helps me to construct commit messages, as you can observe from my commit history - I love to make many small commits.
Theoretically, you can include into the commit hook any actions and auto deliver into the commit message any information you want.

I have a script repository also (https://github.com/Anton-Latukha/fish-functions). There is also interesting stuff for me.

But the Wiegley has the biggest and deservedly famous repositories in this theme.
1. https://github.com/jwiegley/git-scripts (I may taken some from there for me, I do not remember strictly)
2. https://github.com/jwiegley/scripts
3. famous Emacs scripts one
4. as also a lot of git teaching material, so he already knows what we elaborate on.

**Comment by jwiegley** [comment]
Hi @Anton-Latukha, if I may ask, can you please summarize would should be done at this point to close this issue? I've tagged those two commits in `data/nix`, in case that will help, but I'm not sure I've understood what the "right" answer is here.

**Comment by Anton-Latukha** [comment]
@jwiegley 
I've already submitted a PR into haskell-nix/nix#3.

I show the case there.

Since objects are used by people - they need to be in the git branching tree graph for git clone to traverse them.

**Comment by ghost** [comment]
> Hi @Anton-Latukha, if I may ask, can you please summarize would should be done at this point to close this issue? I've tagged those two commits in `data/nix`, in case that will help, but I'm not sure I've understood what the "right" answer is here.

This at least resolves the issues that I had been seeing, where older versions of hnix could not be built. 

This shell command now succeeds, and it failed before setting these tags.
```
nix-shell -p nix-prefetch-git --command "nix-prefetch-git https://github.com/haskell-nix/hnix.git 617d0867ab96c8f97b02c4524bd948d9f114005e --fetch-submodules"
```

I would consider this to resolve this issue, except @Anton-Latukha is still seeing issues with the current state.

Thanks for your help!

**Comment by Anton-Latukha** [comment]
I do not need this, I was trying to help. When I was debugging I found the case, but as we see, that was the other case with the same commits.

For me is important that I understand Git better now, and that you issue is solved. I know now that I need to read on the topic how submodules and tags work internally in Git.

Since you opened this issue and it is resolved - close this issue.
